GetStream/stream-chat-swift
# Add Message Search and Channel Search UI Components

### 🔗 Issue Links
Resolves https://github.com/GetStream/ios-issues-tracking/issues/217

### 🎯 Goal
Add message search and channel search UI Components. 

### 📝 Summary
- Adds Message Search UI Component
- Adds Channel Search UI Component
- Renames `Components.mentionAvatarView` -> `Components.userAvatarView`

### 🛠 Implementation
Two new UI Components were added, `ChatMessageSearchVC` and `ChatChannelSearchVC`. These two new view controllers are then used as a `resultsController` of the `UISearchController` in the Channel List.

Both `ChatMessageSearchVC` and `ChatChannelSearchVC` are subclasses of `ChatChannelListSearchVC`. The latter is an abstract class in which the responsibility is to reuse the common search logic between the Message Search and the Channel Search. In turn, the `ChatChannelListSearchVC` is a subclass of `ChatChannelListVC` since the UI is pretty much the same as the one from the Channel List.

**Class Diagram**
```mermaid
classDiagram
    ChatChannelListSearchVC <|-- ChatChannelSearchVC
    ChatChannelListVC <|-- ChatChannelListSearchVC
    ChatChannelListSearchVC <|-- ChatMessageSearchVC
    class ChatChannelSearchVC{
        loadSearchResults(text: String)*
        loadMoreSearchResults()*
        bool hasEmptyResults*
    }
    class ChatChannelListSearchVC{
        loadSearchResults(text: String)*
        loadMoreSearchResults()*
        bool hasEmptyResults*
    }
    class ChatMessageSearchVC{
        loadSearchResults(text: String)
        loadMoreSearchResults()
        bool hasEmptyResults
    }
```

**Usage:**
By default, the search is disabled. To enable it, a `ChannelListSearchStrategy` should be configured in the `Components` config. The SDK provides two search strategies out-of-the-box, searching by channels or messages. Here is how you can enable the different strategies.
```swift
// If you want to enable the search by channels
Components.default.channelListSearchStrategy = .channels
// If you want to enable the search by messages
Components.default.channelListSearchStrategy = .messages
// If you want to enable the search by channels but with a custom subclass
Components.default.channelListSearchStrategy = .channels(CustomChannelSearchVC.self)
// If you want to enable the search by messages but with a custom subclass
Components.default.channelListSearchStrategy = .messages(CustomMessageSearchVC.self)
```

### 🎨 Showcase
| Channels Search | Messages Search |
|--------|--------|
| <video src="https://github.com/GetStream/stream-chat-swift/assets/12814114/79540469-8d2d-4537-ac32-8966f735e868"/> | <video src="https://github.com/GetStream/stream-chat-swift/assets/12814114/899fa777-28f5-42e0-953e-38f80368dc53"/> | 

### 🧪 Manual Testing Notes
- Open DemoApp
- Tap on Configuration
- Make sure the channel list search strategy is set to messages or channels
- Open the Channel List
- It should have a top search bar
- Search messages or channels
- When tapping a channel, it should open that channel
- When tapping a message, it should open the channel in the message location

**Known backend issue:** If a channel is truncated, the message will still appear on the search results, so when we open the channel on this message, it won't jump to this message. The backend will fix this soon. The expected behaviour is that messages which belong to a truncated channel should not appear in the search result.

### ☑️ Contributor Checklist
